src/stan/lang/grammars/semantic_actions_def.cpp

```diff
@@ -111,6 +112,118 @@ namespace stan {
+                               const scope& var_scope,
+                               const variable_map& vm,
+                               std::ostream& error_msgs) {
+      // validate scope matches declaration scope
+      scope lhs_origin = vm.get_scope(lhs_var_name);
+      if (lhs_origin.program_block() != var_scope.program_block()) {
+        error_msgs << "Cannot assign to variable outside of declaration block"
+                   << "; left-hand-side variable origin=";
+        print_scope(error_msgs, lhs_origin);
+        error_msgs << std::endl;
+        return false;
+      }
+      // enforce constancy of function args
+      if (!lhs_origin.is_local() && lhs_origin.fun()) {
+        error_msgs << "Cannot assign to function argument variables."
+                   << std::endl
+                   << "Use local variables instead."
+                   << std::endl;
+        return false;
+      }
+      return true;
+    }
+
+    expr_type infer_var_dims_type(const base_var_decl& var_type,
+                                  const variable_dims& var_dims) {
+      size_t var_num_dims = var_type.dims_.size();
+      size_t num_index_dims = var_dims.dims_.size();
+      return infer_type_indexing(var_type.base_type_,
+                                 var_num_dims,
+                                 num_index_dims);
+    }
+
+    bool has_same_shape(const expr_type& lhs_type,
+                        const expression& rhs_expr,
+                        const std::string& name,
+                        const std::string& stmt_type,
+                        std::ostream& error_msgs) {
+      if (lhs_type.num_dims_ != rhs_expr.expression_type().num_dims_) {
+        error_msgs << "Dimension mismatch in "
+                   << stmt_type
+                   << "; variable name = "
+                   << name
+                   << ", num dimensions given = "
+                   << lhs_type.num_dims_
+                   << "; right-hand side dimensions = "
+                   << rhs_expr.expression_type().num_dims_
+                   << std::endl;
+        return false;
+      }
+      base_expr_type lhs_base_type = lhs_type.base_type_;
+      base_expr_type rhs_base_type = rhs_expr.expression_type().base_type_;
+      // allow int -> double promotion, even in arrays
+      bool types_compatible =
+        (lhs_base_type == rhs_base_type
+         || (lhs_base_type == DOUBLE_T && rhs_base_type == INT_T));
+      if (!types_compatible) {
+        error_msgs << "Base type mismatch in "
+                   << stmt_type
+                   << "; variable name = "
+                   << name
+                   << ", type = ";
+        write_base_expr_type(error_msgs, lhs_base_type);
+        error_msgs << "; right-hand side type=";
+        write_base_expr_type(error_msgs, rhs_base_type);
+        error_msgs << std::endl;
+        return false;
+      }
+      return true;
+    }
+
+
+    // //////////////////////////////////
```

Comment:
I don't think we need these big, non-standard comments.  But it's not a necessary change.